File: src/main/tray.ts
import { app, Menu, Tray } from "electron";
import icon from '../../resources/nahida.png?asset'
import { mainWindow } from "./window";
import { getAutoUpdater } from "./updater";

function createTray() {
  const tray = new Tray(icon);
  const contextMenu = Menu.buildFromTemplate([
    { label: 'Check Update', type: 'normal', click: async () => { const updater = getAutoUpdater(); await updater.checkForUpdates() } },
    // { type: 'separator' },
    { label: 'Quit', type: 'normal', click: () => { app.quit(); } }
  ]);
  tray.setToolTip('Nahida Desktop');
  tray.setContextMenu(contextMenu);
  tray.on('click', () => {
    mainWindow.show();
  });
  return tray;
}

export { createTray };

File: internal/app/tray.go
package app

import (
	"context"

	"github.com/wailsapp/wails/v3/pkg/application"

	"nahida.live/desktop/internal/infra"
)

func newTray(app *application.App, rt *runtime, icon []byte) *application.SystemTray {
	if app == nil || rt == nil {
		return nil
	}
	tray := app.SystemTray.New()
	if len(icon) > 0 {
		tray.SetIcon(icon)
	}
	tray.SetTooltip("Nahida Desktop")
	menu := app.Menu.New()
	menu.Add("Check for Updates...").OnClick(func(_ *application.Context) {
		if rt.updater == nil {
			return
		}
		go func() {
			if err := rt.updater.CheckForUpdates(context.Background(), true); err != nil && rt.log != nil {
				_ = infra.ReportError(rt.log, err, "updater.manualCheck", infra.Diagnostic{Severity: infra.DiagnosticError, Operation: "updater.manualCheck", Stage: "background"})
			}
		}()
	})
	menu.Add("Setting").OnClick(func(_ *application.Context) {
		if rt.window != nil {
			rt.window.OpenSetting()
		}
	})
	menu.AddSeparator()
	menu.Add("Quit").OnClick(func(_ *application.Context) { app.Quit() })
	tray.SetMenu(menu)
	tray.OnClick(func() {
		if rt.window != nil {
			rt.window.Focus()
		}
	})
	return tray
}
